File: Ex07/Ex07.c
#include <stdio.h>
#include <string.h>

int main() {
    int n;
    scanf("%d", &n);

    static int global;
    int local;

    int *ptr_global = &global;
    int *ptr_local = NULL;

    char comando[50];
    int valor;

    for (int i = 0; i < n; i++) {
        scanf("%s", comando);

        if (strcmp(comando, "SET_GLOBAL") == 0) {
            scanf("%d", &valor);
            global = valor;
        }
        else if (strcmp(comando, "SET_LOCAL") == 0) {
            scanf("%d", &valor);
            local = valor;
            ptr_local = &local;
        }
        else if (strcmp(comando, "READ_GLOBAL") == 0) {
            printf("GLOBAL: %d\n", *ptr_global);
        }
        else if (strcmp(comando, "READ_LOCAL") == 0) {
            if (ptr_local != NULL) {
                printf("LOCAL: %d\n", *ptr_local);
            } else {
                printf("LOCAL: %d\n", *ptr_global);
            }
        }
        else if (strcmp(comando, "RESET_LOCAL") == 0) {
            ptr_local = NULL;
        }
    }

    // saída final
    if (ptr_local != NULL) {
        printf("HISTORICO: G=%d L=%d\n", *ptr_global, *ptr_local);
    } else {
        printf("HISTORICO: G=%d L=NENHUM\n", *ptr_global);
    }

    return 0;
}

File: Ex07/Ex07.py
n = int(input())

escopo = {
    "global": None,
    "local": None
}

def set_global(v):
    global escopo
    escopo['global'] = v

def set_local(v):
    escopo['local'] = v

def read_global():
    print(f"GLOBAL: {escopo['global']}")

def read_local():
    if escopo['local'] is not None:
        print(f"LOCAL: {escopo['local']}")
    else:
        print(f"LOCAL: {escopo['global']}") 

def reset_local():
    escopo['local'] = None

for i in range(n):
    entradaDoUsuario = input()
    partesDoComando = entradaDoUsuario.split()

    comando = partesDoComando[0]

    if comando == "SET_GLOBAL":
        set_global(int(partesDoComando[1]))
    elif comando == "SET_LOCAL":
        set_local(int(partesDoComando[1]))
    elif comando == "READ_GLOBAL":
        read_global()
    elif comando == "READ_LOCAL":
        read_local()
    elif comando == "RESET_LOCAL":
        reset_local()

# saída final
if escopo["local"] is not None:
    print(f"HISTORICO: G={escopo['global']} L={escopo['local']}")
else:
    print(f"HISTORICO: G={escopo['global']} L=NENHUM")
